drivers/mtd/mtd_config_fs.c

```diff
@@ -0,0 +1,2654 @@
+{
+  uint8_t write_block_size = NVS_WRITE_BLOCK_SIZE;
+
+  if (write_block_size <= 1U)
+    {
+      return len;
+    }
+
+  return (len + (write_block_size - 1U)) & ~(write_block_size - 1U);
+}
+
+/****************************************************************************
+ * Name: nvs_flash_al_wrt
+ *
+ * Description:
+ *   flash routines, basic aligned flash write to nvs address.
+ *
+ ****************************************************************************/
+
+static int nvs_flash_al_wrt(FAR struct nvs_fs *fs, uint32_t addr,
+                             FAR const void *data, size_t len)
+{
+  FAR const uint8_t *data8 = (FAR const uint8_t *)data;
+  int rc = 0;
+  off_t offset;
+  size_t blen;
+  uint8_t buf[NVS_BLOCK_SIZE];
+
+  if (!len)
+    {
+      /* Nothing to write, avoid changing the flash protection */
+
+      return 0;
+    }
+
+  offset = fs->sector_size * (addr >> ADDR_SECT_SHIFT);
+  offset += addr & ADDR_OFFS_MASK;
+
+  blen = len & ~(NVS_WRITE_BLOCK_SIZE - 1U);
+  if (blen > 0)
+    {
+      rc = flash_write(fs->mtd, offset, data8, blen);
+      if (rc)
+        {
+          /* flash write error */
+
+          goto end;
+        }
+
+      len -= blen;
+      offset += blen;
+      data8 += blen;
+    }
+
+  if (len)
+    {
+      memcpy(buf, data8, len);
+      (void)memset(buf + len, CONFIG_MTD_CONFIG_ERASEDVALUE,
+                   NVS_WRITE_BLOCK_SIZE - len);
+
+      rc = flash_write(fs->mtd, offset, buf,
+                       NVS_WRITE_BLOCK_SIZE);
+      if (rc)
+        {
+          /* flash write error */
+
+          goto end;
+        }
+    }
+
+end:
+  return rc;
+}
+
+/****************************************************************************
+ * Name: nvs_flash_rd
+ *
+ * Description:
+ *   basic flash read from nvs address.
+ *
+ ****************************************************************************/
+
+static int nvs_flash_rd(FAR struct nvs_fs *fs, uint32_t addr,
+                        FAR void *data, size_t len)
+{
+  int rc;
+  off_t offset;
+
+  offset = fs->sector_size * (addr >> ADDR_SECT_SHIFT);
+  offset += addr & ADDR_OFFS_MASK;
+
+  rc = flash_read(fs->mtd, offset, data, len);
+  return rc;
+}
+
+/****************************************************************************
+ * Name: nvs_flash_ate_wrt
+ *
+ * Description:
+ *   allocation entry write.
+ *
+ ****************************************************************************/
+
+static int nvs_flash_ate_wrt(FAR struct nvs_fs *fs,
+                                FAR const struct nvs_ate *entry)
+{
+  int rc;
+
+  rc = nvs_flash_al_wrt(fs, fs->ate_wra, entry, sizeof (struct nvs_ate));
+
+#ifdef CONFIG_MTD_CONFIG_FAIL_SAFE_LOOKUP_CACHE
+  /* 0xffffffff is a special-purpose identifier. Exclude it from the cache */
+
+  if (entry->id != 0xffffffff)
+    {
+      fs->lookup_cache[nvs_lookup_cache_pos(entry->id)] = fs->ate_wra;
+    }
+#endif
+
+  fs->ate_wra -= nvs_al_size(fs, sizeof(struct nvs_ate));
```

Comment:
do we need to check `rc` in prior to this actions?

Reply:
The original zephyr implementation is exactly the same. I guess it tolerates some flash write  error, and moves on.
So I'd rather not change its behavior.